### PR TITLE
igr-tests: Refactoring

### PR DESCRIPTION
**Pull request by mobin-2008**
Hi!
Dinit currently use a simple runner for running some scripts as "integration tests". but it system needs refactoring.
> Why?

Because it's have many problems and bugs [Highest priority to lowest]:
1. Tests who use dinit socket may hangs forever. look at this example:
     https://github.com/davmac314/dinit/blob/b9f20c58a23a50493d2cea07985cadb25381de5f/src/igr-tests/force-stop/run-test.sh#L5-L11
    Current igr-tests doesn't have `set -e` so if a command fails it doesn't exit immediately so it hangs forever and need manual cancellation. Of course if command runs successfully but refuse to create socket, it hangs again. I think it's a bug due to poor design.

2. Many statements don't have error/warning messages. error messages are very useful in many statements, such as when output file isn't exist, It helps developers to faster debugging.

3. Many things are repeated in tests, Things who can be in a reference file (such as launching dinit, stopping dinit and ...).

4. Current implementation isn't Meson compatible. For meson builds we just copy entire igr-tests into build environment and run tests. I think that's can be changed if we have something like dynamic `$TEMP`.

5. And many other things such as all igr-tests do same tasks via different ways without any reason (some scripts disable dinit via `kill` and some other via `dinitctl` and ...)

I think it's definitely reasonable.

Regards.

# ToDo
- [x] Create initial `igr_functions`
- [ ] Refactor `compare_*()` functions (Or even remove them)
- [ ] support `DEBUG` keyword for avoid `--quiet` in dinit commands.
- [ ] `set -e` for all tests
  - [ ] I'm not sure about `set -u`
- [ ] Refactor all scripts
- [ ] Enable tests for meson
  - [ ] Fix `TEST_NAME` variable
- [ ] Complete docs

`Signed-off-by: Mobin "Hojjat" Aydinfar <mobin@mobintestserver.ir>`

### CODE REVIEW

**Comment by davmac314** [comment]
I'm unsubscribing from notifications, let me know when this is ready for review.

By the way - opening a draft PR always notifies me, and notifies me every commit, unless I specifically unsubscribe. If possible can you avoid opening PRs until they are actually ready? It is a pain but there seems to be no way I can avoid getting notifications for draft PRs on Github.

**Comment by davmac314** [comment]
> Hmm, From https://github.blog/2019-02-14-introducing-draft-pull-requests/:
> 
> > Also, if you have a [CODEOWNERS file](https://github.blog/2017-07-06-introducing-code-owners/) in your repository, a draft pull request will suppress notifications to those reviewers until it is marked as ready for review.

Hmm I could try this but I suspect it will still notify me because I am repository owner (whether "code owner" or not).
 
> OK, I just open issue for in-progress branches.

It's up to you - you can open draft PRs if you prefer, but I will need to unsubscribe from them so I will only see messages and status changes if you specifically mention me via @davmac314

File: src/igr-tests/igr-runner.cc [hunk 1]
```diff
@@ -14,12 +14,12 @@ int main(int argc, char **argv)
...
-            "no-command-error", "add-rm-dep", "var-subst", "svc-start-fail", "dep-not-found", "pseudo-cycle" };
+            /* "no-command-error", "add-rm-dep", "var-subst", "svc-start-fail", "dep-not-found", "pseudo-cycle" */};
     constexpr int num_tests = sizeof(test_dirs) / sizeof(test_dirs[0]);
 
-    int passed = 0;
-    int skipped = 0;
-    int failed = 0;
+    unsigned int passed = 0;
+    unsigned int skipped = 0;
+    unsigned int failed = 0;
```
Comment (mobin-2008) [inline src/igr-tests/igr-runner.cc:22]: @davmac314 I think it's must be `unsigned short int` for better memory management. No?

Comment (davmac314) [reply src/igr-tests/igr-runner.cc:22]: Changing these to `short` would literally save 6 bytes, while running the integration tests only. I really have better things to worry about!

**Comment by mobin-2008** [comment]
Seems like branch was disconnected from this PR. (PR didn't updated) but seems like I can't reopen it due to rebased branch :/

**Comment by mobin-2008** [comment]
-

**Comment by davmac314** [comment]
Strange...

I can review it (and merge it) manually if you continue working in the branch. Otherwise try making a new branch from this one and open a new PR maybe.

**Comment by mobin-2008** [comment]
> Strange...
> 
> I can review it (and merge it) manually if you continue working in the branch. Otherwise try making a new branch from this one and open a new PR maybe.

Yea, I working on refactor_igr_tests branch and it reaches a good state (tests are refactored except environ2 due to some reasons). I will open PR if most things become ready.

**Pull request closed without merging**